engine/common/synchronization/finalized_snapshot.go

```diff
@@ -0,0 +1,125 @@
+package synchronization
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/lifecycle"
+	"github.com/onflow/flow-go/state/protocol"
+)
+
+// finalizedSnapshot is a helper structure which contains latest finalized header and participants list
+type finalizedSnapshot struct {
+	head         *flow.Header
+	participants flow.IdentityList
+}
+
+// FinalizedSnapshotCache represents a cached snapshot of the latest finalized header and participants list.
+// It is used in Engine to access latest valid data.
+type FinalizedSnapshotCache struct {
+	mu sync.RWMutex
+
+	log                       zerolog.Logger
+	state                     protocol.State
+	identityFilter            flow.IdentityFilter
+	lastFinalizedSnapshot     *finalizedSnapshot
+	finalizationEventNotifier engine.Notifier // notifier for finalization events
+
+	lm *lifecycle.LifecycleManager
+}
+
+// NewFinalizedSnapshotCache creates a new finalized snapshot cache.
+func NewFinalizedSnapshotCache(log zerolog.Logger, state protocol.State, participantsFilter flow.IdentityFilter) (*FinalizedSnapshotCache, error) {
+
+	cache := &FinalizedSnapshotCache{
+		state:          state,
+		identityFilter: participantsFilter,
+		lm:             lifecycle.NewLifecycleManager(),
+		log:            log.With().Str("component", "finalized_snapshot_cache").Logger(),
+	}
+
+	err := cache.updateSnapshot()
+	if err != nil {
+		return nil, fmt.Errorf("could not apply last finalized state")
+	}
+
+	return cache, nil
+}
+
+// get returns last locally stored snapshot which contains final header
+// and list of filtered identities
+func (f *FinalizedSnapshotCache) get() *finalizedSnapshot {
+	f.mu.RLock()
+	defer f.mu.RUnlock()
+	return f.lastFinalizedSnapshot
+}
+
+// updateSnapshot updates latest locally cached finalized snapshot
+func (f *FinalizedSnapshotCache) updateSnapshot() error {
+	finalSnapshot := f.state.Final()
+	head, err := finalSnapshot.Head()
+	if err != nil {
+		return fmt.Errorf("could not get last finalized header: %w", err)
+	}
+
+	// get all participant nodes from the state
+	participants, err := finalSnapshot.Identities(f.identityFilter)
+	if err != nil {
+		return fmt.Errorf("could get consensus participants at latest finalized block: %w", err)
+	}
+
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	if f.lastFinalizedSnapshot != nil && f.lastFinalizedSnapshot.head.Height >= head.Height {
+		return nil
+	}
+
+	f.lastFinalizedSnapshot = &finalizedSnapshot{
+		head:         head,
+		participants: participants,
```

Comment:
Thanks for pointing this out, I'll take a look at this.

Reply:
@zhangchiqing it may actually be [2 seconds](https://github.com/onflow/flow-go/blob/master/engine/common/synchronization/config.go#L15) since we also access it [here](https://github.com/onflow/flow-go/blob/smnzhu%2Fsync-request-handler/engine/common/synchronization/engine.go#L326-L329). Nevertheless, the [block finalization rate](https://dapperlabs.grafana.net/d/PkvVJj4Mz/mainnet-general?orgId=1&refresh=10s&viewPanel=4) is around 2 blocks per second, so I think you're right that We can just remove this field.

Reply:
It was my invention to put participants list into snapshot to reduce locking of `state`. I believe for synchronization engine it's ok to fetch participants list in advance since it's used at least in two places. So I wouldn't remove it.

Reply:
d80fdd7

Reply:
@durkmurder I get the participants list based on the block ID of the cached finalized header, so I don't think this causes any extra locking.

Take a look at the last two commits and lmk what you think.